test/e2e/inspection_test.go

```diff
@@ -0,0 +1,155 @@
+			Creator:   clusterProxy.GetClient(),
+			ClientSet: clusterProxy.GetClientSet(),
+			Name:      fmt.Sprintf("%s-%s", specName, util.RandomString(6)),
+			LogFolder: artifactFolder,
+		})
+	})
+
+	It("should put BMH without BMC credentials in unmanaged state", func() {
+		By("creating a BMH")
+		bmh := metal3api.BareMetalHost{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      specName + "-unmanaged",
+				Namespace: namespace.Name,
+			},
+		}
+		err := clusterProxy.GetClient().Create(ctx, &bmh)
+		Expect(err).NotTo(HaveOccurred())
+
+		By("waiting for the BMH to be in unmanaged state")
+		WaitForBmhInState(ctx, WaitForBmhInStateInput{
+			Client: clusterProxy.GetClient(),
+			Bmh:    bmh,
+			State:  metal3api.StateUnmanaged,
+		}, e2eConfig.GetIntervals(specName, "wait-unmanaged")...)
+	})
+
+	It("should fail to register the BMH if the secret is missing", func() {
+		By("creating a BMH")
+		bmh := metal3api.BareMetalHost{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      specName + "-registration-error",
+				Namespace: namespace.Name,
+			},
+			Spec: metal3api.BareMetalHostSpec{
+				BMC: metal3api.BMCDetails{
+					Address:         "ipmi://127.0.0.1:5678",
+					CredentialsName: "bmc-credentials",
+				},
+			},
+		}
+		err := clusterProxy.GetClient().Create(ctx, &bmh)
+		Expect(err).NotTo(HaveOccurred())
+
+		By("trying to register the BMH")
+		WaitForBmhInState(ctx, WaitForBmhInStateInput{
+			Client: clusterProxy.GetClient(),
+			Bmh:    bmh,
+			State:  metal3api.StateRegistering,
+		}, e2eConfig.GetIntervals(specName, "wait-registering")...)
+
+		By("waiting for registration error on the BMH")
+		Eventually(func(g Gomega) {
+			key := types.NamespacedName{Namespace: bmh.Namespace, Name: bmh.Name}
+			g.Expect(clusterProxy.GetClient().Get(ctx, key, &bmh)).To(Succeed())
+			g.Expect(bmh.Status.ErrorType).To(Equal(metal3api.RegistrationError))
+		}, e2eConfig.GetIntervals(specName, "wait-registration-error")...).Should(Succeed())
+	})
+
+	It("should inspect a newly created BMH", func() {
+		By("creating a secret with BMH credentials")
+		bmcCredentials := corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "bmc-credentials",
+				Namespace: namespace.Name,
+			},
+			StringData: map[string]string{
+				"username": bmcUser,
+				"password": bmcPassword,
+			},
+		}
+		err := clusterProxy.GetClient().Create(ctx, &bmcCredentials)
+		Expect(err).NotTo(HaveOccurred())
+
+		By("creating a BMH")
+		bmh := metal3api.BareMetalHost{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      specName + "-inspect",
+				Namespace: namespace.Name,
+			},
+			Spec: metal3api.BareMetalHostSpec{
+				BMC: metal3api.BMCDetails{
+					Address:         bmcAddress,
+					CredentialsName: "bmc-credentials",
+				},
+				BootMode:       metal3api.Legacy,
+				BootMACAddress: bootMacAddress,
+			},
+		}
+		err = clusterProxy.GetClient().Create(ctx, &bmh)
+		Expect(err).NotTo(HaveOccurred())
+
+		By("waiting for the BMH to be in registering state")
```

Comment:
Is there a risk that this state will happen too quickly, causing flakes in the CI?

Reply:
Potentially yes, but I have not seen a single flake so far. Ginkgo provides a really nice way to deal with this by taking in a deadline and check interval. I exposed these through the config file in the same way as CAPI is doing, which allows us to have separate config for separate environments if we want.

To see how well this works, check the `ironic.yaml` and `fixture.yaml`. Thanks to the config we can run the same test against BMO in test mode with millisecond checks or agains BMO + Ironic with 5 seconds between checks and a deadline that is several minutes.

Intervals for BMO + Ironic (the first value is the deadline, second is how often it is checked):
https://github.com/metal3-io/baremetal-operator/blob/dd2d2bd10d11c4da74ec6ab79a5376ce2f93f5fa/test/e2e/config/ironic.yaml#L21-L28

Intervals for BMO fixture:
https://github.com/metal3-io/baremetal-operator/blob/dd2d2bd10d11c4da74ec6ab79a5376ce2f93f5fa/test/e2e/config/fixture.yaml#L18-L25

Reply:
Edit: I have of course seen flakes while playing with these values, but after finding values that works there has been no issues :slightly_smiling_face: